memstore/list/live_vector_party.go

```diff
@@ -0,0 +1,394 @@
+
+package list
+
+import (
+	"github.com/uber/aresdb/cgoutils"
+	"github.com/uber/aresdb/memstore"
+	"github.com/uber/aresdb/memstore/common"
+	"github.com/uber/aresdb/utils"
+	"io"
+	"sync"
+	"unsafe"
+)
+
+// LiveVectorParty is the representation of list data type vector party in live store.
+// It supports random access read and write. However, it does not support serialization into disk.
+// It underlying uses a high level memory pool to store the list data. Therefore when this vector
+// party is destructed, the underlying memory pool needs to be destroyed as well.
+type LiveVectorParty struct {
+	baseVectorParty
+	// storing the offset to slab footer offset for each row.
+	caps       *memstore.Vector
+	memoryPool HighLevelMemoryPool
+	sync.RWMutex
+}
+
+// GetBytes returns the bytes this vp occupies.
+func (vp *LiveVectorParty) GetBytes() int64 {
+	vp.RLock()
+	defer vp.RUnlock()
+
+	var bytes int64
+	if vp.memoryPool != nil {
+		bytes += vp.memoryPool.GetNativeMemoryAllocator().GetTotalBytes()
+	}
+
+	if vp.offsets != nil {
+		bytes += int64(vp.offsets.Bytes)
+	}
+	if vp.caps != nil {
+		bytes += int64(vp.caps.Bytes)
+	}
+	return bytes
+}
+
+// SafeDestruct destructs vector party memory.
+func (vp *LiveVectorParty) SafeDestruct() {
+	if vp != nil {
+		totalBytes := vp.GetBytes()
+		if vp.offsets != nil {
+			vp.offsets.SafeDestruct()
+			vp.offsets = nil
+		}
+		if vp.caps != nil {
+			vp.caps.SafeDestruct()
+			vp.caps = nil
+		}
+		if vp.memoryPool != nil {
+			vp.memoryPool.Destroy()
+			vp.memoryPool = nil
+		}
+		if totalBytes > 0 {
+			vp.reporter(-totalBytes)
+		}
+	}
+}
+
+// Write serialize vector party.
+func (vp *LiveVectorParty) Write(writer io.Writer) (err error) {
+	vp.RLock()
+	defer vp.RUnlock()
+
+	dataWriter := utils.NewStreamDataWriter(writer)
+	if err = dataWriter.WriteUint32(ListVectorPartyHeader); err != nil {
+		return
+	}
+
+	// length
+	if err = dataWriter.WriteInt32(int32(vp.length)); err != nil {
+		return
+	}
+
+	// data type
+	if err = dataWriter.WriteUint32(uint32(vp.dataType)); err != nil {
+		return
+	}
+
+	// nonDefaultValue count, 0 for List VectorParty
+	if err = dataWriter.WriteInt32(int32(0)); err != nil {
+		return
+	}
+
+	// columnMode, AllValuesPresent for now
+	columnMode := common.AllValuesPresent
+	if err = dataWriter.WriteUint16(uint16(columnMode)); err != nil {
+		return
+	}
+
+	// Write 6 bytes padding.
+	if err = dataWriter.SkipBytes(6); err != nil {
+		return
+	}
+
+	// write offsets
+	var valueBytes int
+	for i := 0; i < vp.length; i++ {
+		_, length := vp.GetOffsetLength(i)
+		if length == 0 {
+			dataWriter.WriteUint32(uint32(0))
+		} else {
+			dataWriter.WriteUint32(uint32(valueBytes))
+		}
+		dataWriter.WriteUint32(length)
+
+		valueBytes += common.CalculateListElementBytes(vp.dataType, int(length))
+	}
+	// to compatible with archive vp, align to 64 bytes alignment
+	if vp.offsets.Bytes > 8*vp.length {
+		dataWriter.SkipBytes(vp.offsets.Bytes - 8*vp.length)
```

Comment:
there is a call called WritePadding we can use

Reply:
good one, updated